README.md
<table style="margin:0px auto; border:0px">
	<tr style="text-align:center; border:0px">
		<td style="text-align:center; padding:10px 20px;
			border:2px steelblue solid; border-radius:25px;">
		<a href="https://socratic-software.github.io/stochastica/">
		<img src="images/StochasticA.png" height="30%" style="margin-bottom:15px;">
		<br><span style="font-family:Arial, Helvetica, sans-serif; font-size:80%;">
		StochasticA<br>26 Laboratory Experiments</span></a></td>
		<td>&nbsp;&nbsp;&nbsp;&nbsp;</td>
		<td style="text-align:center; padding:10px 20px;
			border:2px steelblue solid; border-radius:25px;">
		<a href="https://apps.apple.com/us/app/stochastic-signal-processing/id1450268179">
		<img src="images/StochasticS.png" height="30%" style="margin-bottom:15px;">
		<br><span style="font-family:Arial, Helvetica, sans-serif; font-size:80%;">
		StochasticS (iOS)<br>59 Laboratory Experiments</span></a></td>
		<td>&nbsp;&nbsp;&nbsp;&nbsp;</td>
		<td style="text-align:center; padding:10px 20px;
			border:2px steelblue solid; border-radius:25px;">
		<a href="https://play.google.com/store/apps/details?id=org.SocraticSoftware.iSSP">
		<img src="images/StochasticS.png" height="30%" style="margin-bottom:15px;">
		<br><span style="font-family:Arial, Helvetica, sans-serif; font-size:80%;">
		StochasticS (Android)<br>59 Laboratory Experiments</span></a></td>
	</tr>
</table>

# StochasticA

## Background
This repository contains the *iBook* / website for *Introduction to Stochastic Signal Processing*. The short name is *StochasticA*.

Speech, music, seismic vibrations, oil prices, and climate measurements are all examples of stochastic (random) signals. In this textbook, intended for individuals with prior training in introductory signal processing and introductory probability theory, we develop techniques to process such signals to extract useful information. We present case studies ranging from music to photographic images to oil prices to climate data to the motion of individual biomolecules.
<br><br>
This textbook, as an interactive textbook (*iBook*), makes use of your device's ability to display dynamic information through images, films, and animations and to hear the results of the techniques applied to music. At the end of every chapter there are homework problems ranging from easy to "olympic".
<br><br>
A new and exciting aspect is that we make use of your device's interactive capabilities to offer 26 laboratory experiments in signal processing. For the web-based implementation in this repository, the experiments use the speakers, display, and the graphical user interface (GUI). For a tablet and smartphone implementation that can be found [here](https://apps.apple.com/us/app/stochastic-signal-processing/id1450268179?ls=1), this is extended to use the camera and the microphone providing a total of 59 laboratory experiments.
<br><br>
These experiments are not simulations; they are examples of real digital processing of signals in your device. In this time of at-home and online learning, this is the way to learn signal processing through study and experimentation.

## IPR
Unless otherwise specified, *html* files are copyrighted and do not fall under the MIT License. Javascript (*js*) files do, in general, fall under the MIT License unless explicitly stated otherwise.

## Privacy Policy
Our *Privacy Policy* can be found [here](https://sites.google.com/socraticsoftware.org/socraticsoftware/privacy-policy)

## Chapter Contents

1. How to use this iBook
1. Prologue
1. Introduction
1. Characterization of Random Signals
1. Correlations and Spectra
1. Filtering of Stochastic Signals
1. The Langevin Equation &mdash; A Case Study
1. Characterizing Signal-to-Noise Ratios
1. The Matched Filter
1. The Wiener filter
1. Aspects of Estimation
1. Spectral Estimation
1. Appendices
1. Information


*Last revision this document: 25 February 2022*


ChangeLog.md
# Changelog

All notable changes to **<font color="steelblue">StochasticA</font>** will be documented in this file.

## [2.4.2] - 2022-02-225

### Updated

- The text describing the Schrödinger’s cat problem (4.2) has been improved.
- Maintenance updates of various libraries. Nothing to see here folks.
- README file.

## [2.4.1] - 2021-12-28

### Added

- Alerts in various Laboratory Exercises when computational overflows or underflows occur.

### Updated

- Maintenance, formatting, and cosmetic improvements.
- README file.

### Fixed

- A small bug in 1-D Fourier calculations when data length is not a power of two..

## [2.4.0] - 2021-06-11

### Changed

- Where appropriate, length of audio signals changed to 2^N (e.g. N=16 or 17) so that processing speed is improved. This is important for slower platforms.

### Updated

- Maintenance, formatting, and cosmetic improvements.
- README file.

## [2.3.6] - 2021-05-17

### Added

- Timing of computation to Laboratory Exercise 6.2.

### Updated

- README file.

## [2.3.5] - 2021-04-26

### Removed

- Extraneous alerts in various Laboratory Exercises.

### Fixed

- Potential bugs and spelling errors in Laboratory Exercises in Chapter 6.

## [2.3.4] - 2021-04-19

### Improved

- Speed in Laboratory Exercises 4.2 and 11.1.
- Formatting in Laboratory Exercises to adjust for user platform and graphical interaction.

### Fixed

- Bug in Laboratory Exercise 6.6.

## [2.3.3] - 2021-04-12

### Changed

- Line spacing in Laboratory Exercise descriptions to adjust for user platform.

## [2.3.2] - 2021-04-12

### Changed

- Font sizes in Laboratory Exercise descriptions to adjust for user platform.

## [2.3.1] - 2021-04-11

### Changed

- Formatting of "radio buttons" to be compatible with varying browser window formats and platforms.

## [2.3.0] - 2021-04-10

### Added

- Tip in chapter info.html to describe how to successfully refresh a page of the textbook or a laboratory exercise (as needed).
- Latest version of plotly.js (v.2.0.0-rc.0 @ github.com/plotly/plotly.js/releases).

### Changed

- Correction of an equation display in a footnote in Chapter 9.
- Formatting of graph displays to be compatible with varying browser window formats.
- Formatting of font sizes in text and graphs to be compatible with varying browser window formats.
- Many other minor textual improvements and corrections as well as maintenance and and cosmetic improvements.

## [2.2.13] - 2021-01-25

### Added

- Latest version of plotly.js (v.1.58.4 @ github.com/plotly/plotly.js/releases).

### Changed

- Corrections to Chapter 10.
- info.html
- README.md

## [2.2.12] - 2020-11-29

### Changed

- Corrected version number

## [2.2.11] - 2020-11-27

### Changed

- Minor textual change in Chapter 1
- Minor textual correction in Chapter 3

## [2.2.11] - 2020-11-18

### Added

- Google analytics added to Chapters and Laboratory Exercises.

### Changed

- Correction to shortcut icon path for favicon.png
- Minor textual correction in Chapter 3

### Removed

- Extraneous UTF-8 in <head>

## [2.2.11] - 2020-11-16

### Added

- Tip in info.html.

## [2.x.y] - 2020-11-07

### Added

- Initial commit of **StochasticA**. Version numbers started with 2.0.0 but have been lost in the mists of time.



js/SSPconstants.js
// ****************************************************************
// Copyright I.T. Young & R. Ligteringen, 1984, 2010–2022
// Licensed under the MIT license.
// ****************************************************************
// SSPconstants.js
//
// Math and signal processing constants to be used in the SSP package
// i.t. young
// Tuesday, 5 July 2016
// Tuesday, 13 December 2016
// Monday, 29 May 2017
// Tuesday, 27 June 2017
// Wednesday, 28 June 2017
// Tuesday, 4 July 2017
// Sunday, 13 August 2017
// Tuesday, 18 September 2018
// 

// The "universal" signal format is characterized by an array name (e.g. "data"),
// an array "nn", and a number of dimensions "ndim". The "data" array takes data
// in the form of [r1,i1,r2,i2, ...,rn,in] where "r1" is the real part of a
// complex sample and "i1" is the imaginary part.
// 
// For a 1D signal, "nn" is an array with one element that gives the number of
// complex signal samples (e.g. nn[0] = 17). The number of dimensions
// ndim = nn.length. In this example ndim = 1. Note that because the data are
// complex data.length = 2*17 = 34. Signals that are 1D can have ANY length.

// For a 2D signal (image), "nn" is an array with two elements that gives
// the number of rows and columns where each sample is complex, that is,
// nn = [rows,columns]. In this case, the number of dimensions
// ndim = nn.length = 2. Obviously, rows - n[0] and columns = nn[1].

// For 2D signals there is a restriction. The number of rows MUST be a power
// of two, e.g. rows = 2^7 and the number of columns MUST also be a power of
// two, e,g.columns =  2^8. (See the routine "fourND" in "SSPfourier.js" for
// an explanation.). This means the data.length = 2*rows*columns = 2*(2^7)*(2^8)
// = 2^16. Caveat emptor!

// Added a formatting subroutine to be used with plotly.js. This seemed like
// the best place as it is not signal processing


// ****************************************************************

var testMe = true; /* default */
var rounding = 0.001;
var spaces5 = '     ';  // how to fool CSS, HTML, javascript, etc.

const FORWARD = 1;
const BACKWARD = -1;

const CORRELATION = true;
const CONVOLUTION = false;

// Digital filters
var leftCoeffs = [];
var rightCoeffs = [];
var dummyFilter = [ leftCoeffs, rightCoeffs ];

// Global constants
const tooSmall = 1e-10;
const tooLarge = 1e+10;
const perc1 = 0.01;
const perc5 = 0.05;
const perc10 = 0.10;
const epsilon = 0.0001;
const canvasDelay = 500;  // milliseconds
const plotlyDelay = 2*canvasDelay;  // milliseconds

const iTMAX = 100;    // Numerical Recipes gamma functions
const ePS = 3e-07;    // Numerical Recipes gamma functions
const fPMIN = 1e-30;  // Numerical Recipes gamma functions

// For the sampling frequency, see: 
// <https://documentation.apple.com/en/soundtrackpro/usermanual/index.html#chapter=B%26section=2%26tasks=true>
//
const sampFreq = 44100;  // Hz = (2*3*5*7)^2
const timeWindow = 3.0;  // max length of time signal in [s]
const factor2 = 17; // audio signals will frequently be this long ≈ 3 seconds

var lengthMax = timeWindow*sampFreq;  // 3.0s x 44.1 kHz sampling rate

const maxAudioRecording = 5;  // seconds

// Musical Notes - See <https://pages.mtu.edu/~suits/notefreqs.html>
const notefactor = 2**(1/12);		// equal-tempered scale

// the base note frequency
const A4 = 440;						// 440 Hz
// other note frequencies
const A4sharp = A4*notefactor;		// 466.16 Hz
const B4flat = A4sharp;				// 466.16 Hz
const B4 = B4flat*notefactor;		// 493.88 Hz
const C5 = B4*notefactor;			// 523.25 Hz
const C5sharp = C5*notefactor;		// 554.37 Hz
const D5flat = C5sharp;				// 554.37 Hz
const D5 = D5flat*notefactor;		// 587.33 Hz
const D5sharp = D5*notefactor;		// 622.25 Hz
const E5flat = D5sharp;				// 622.25 Hz
const E5 = E5flat*notefactor;		// 659.25 Hz
const F5 = E5*notefactor;			// 698.46 Hz
const F5sharp = F5*notefactor;		// 739.99 Hz
const G5flat = F5sharp;				// 739.99 Hz
const G5 = G5flat*notefactor;		// 783.99 Hz
const G5sharp = G5*notefactor;		// 830.61 Hz
const A5flat = G5sharp;				// 830.61 Hz
const C6 = 2*C5;					// 1046.50 Hz

const B3 = B4/2;					// 246.94 Hz
const E4 = E5/2;					// 329.63 Hz
const B5 = B4*2;					// 987.77 Hz
const A3flat = (A5flat/4);			// 207.65 Hz
const A3 = A4/2;					// 880 Hz Hz
const A5 = A4*2;					// 880 Hz Hz
const G3 = G5/4;					// 196.00 Hz

// codes for use in color processing
const redI = 0;
const greenI = 1;
const blueI = 2;


// ****************************************************************

